src/coreclr/utilcode/executableallocator.cpp

```diff
@@ -0,0 +1,755 @@
+    }
+
+    // Fall through to
+#endif // USE_UPPER_ADDRESS
+
+    if (result == NULL)
+    {
+        if (IsDoubleMappingEnabled())
+        {
+            CRITSEC_Holder csh(m_CriticalSection);
+
+            bool isFreeBlock;
+            BlockRX* block = AllocateBlock(size, &isFreeBlock);
+            if (block == NULL)
+            {
+                return NULL;
+            }
+
+            result = (BYTE*)VMToOSInterface::ReserveDoubleMappedMemory(m_doubleMemoryMapperHandle, block->offset, size, 0, 0);
+
+            if (result != NULL)
+            {
+                block->baseRX = result;
+                AddRXBlock(block);
+            }
+            else 
+            {
+                BackoutBlock(block, isFreeBlock);
+            }
+        }
+        else
+        {
+            DWORD allocationType = MEM_RESERVE;
+#ifdef HOST_UNIX
+            // Tell PAL to use the executable memory allocator to satisfy this request for virtual memory.
+            // This will allow us to place JIT'ed code close to the coreclr library
+            // and thus improve performance by avoiding jump stubs in managed code.
+            allocationType |= MEM_RESERVE_EXECUTABLE;
+#endif
+            result = (BYTE*)ClrVirtualAlloc(NULL, size, allocationType, PAGE_NOACCESS);
+        }
+    }
+
+    return result;
+}
+
+// Reserve a block of executable memory at the specified virtual address. If it is not
+// possible, the method returns NULL.
+void* ExecutableAllocator::ReserveAt(void* baseAddressRX, size_t size)
+{
+    LIMITED_METHOD_CONTRACT;
+
+    _ASSERTE((size & (Granularity() - 1)) == 0);
+
+    if (IsDoubleMappingEnabled())
+    {
+        CRITSEC_Holder csh(m_CriticalSection);
+
+        bool isFreeBlock;
+        BlockRX* block = AllocateBlock(size, &isFreeBlock);
+        if (block == NULL)
+        {
+            return NULL;
+        }
+
+        void* result = VMToOSInterface::ReserveDoubleMappedMemory(m_doubleMemoryMapperHandle, block->offset, size, baseAddressRX, baseAddressRX);
+
+        if (result != NULL)
+        {
+            block->baseRX = result;
+            AddRXBlock(block);
+        }
+        else 
+        {
+            BackoutBlock(block, isFreeBlock);
+        }
+
+        return result;
+    }
+    else
+    {
+        return VirtualAlloc(baseAddressRX, size, MEM_RESERVE, PAGE_NOACCESS);
+    }
+}
+
+// Map an executable memory block as writeable. If there is already a mapping
+// covering the specified block, return that mapping instead of creating a new one.
+// Return starting address of the writeable mapping.
+void* ExecutableAllocator::MapRW(void* pRX, size_t size)
+{
+    LIMITED_METHOD_CONTRACT;
+
+    if (!IsDoubleMappingEnabled())
+    {
+        return pRX;
+    }
+
+    CRITSEC_Holder csh(m_CriticalSection);
+
+    void* result = FindRWBlock(pRX, size);
+    if (result != NULL)
+    {
+        return result;
+    }
+
+    for (BlockRX* pBlock = m_pFirstBlockRX; pBlock != NULL; pBlock = pBlock->next)
```

Comment:
This would likely be worth optimizing using some better data structure than a linked list to better handle apps that JIT a lot of code. Testing this with aspnet plaintext benchmark, the max length of this list was about 100 and the average search steps to find the requested block was 27. The list has the most recent allocated block in the head, so for the case of allocating memory block and then filling it with code, the current way would perform the best as the block we try to map as rw would always be the first in the list. But for modifications of old allocated stuff, like patching stubs etc., we can end up doing many search steps.
However, the plan for post 6.0 is to try to eliminate as much of the patching of code as we can e.g. by splitting stubs into fixed code and writeable data, so I didn't want to optimize this prematurely.